gpu/src/fft/fft_metal.rs

```diff
@@ -0,0 +1,257 @@
+use std::mem;
+
+use lambdaworks_math::field::{element::FieldElement, traits::IsTwoAdicField};
+
+use super::helpers::void_ptr;
+use lambdaworks_math::fft::errors::FFTError;
+use metal::{CommandQueue, Device, Library, MTLResourceOptions, MTLSize};
+
+const FFT_LIB_DATA: &[u8] = include_bytes!("../metal/fft.metallib");
+
+pub struct FFTMetalState {
+    pub device: Device,
+    pub library: Library,
+    pub command_queue: CommandQueue,
+}
+
+impl FFTMetalState {
+    pub fn new(device: Option<Device>) -> Result<Self, FFTError> {
+        let device: Device =
+            device.unwrap_or(Device::system_default().ok_or(FFTError::MetalDeviceNotFound())?);
+
+        let library = device
+            .new_library_with_data(FFT_LIB_DATA)
+            .map_err(FFTError::MetalLibraryError)?;
+        let command_queue = device.new_command_queue();
+
+        Ok(FFTMetalState {
+            device,
+            library,
+            command_queue,
+        })
+    }
+
+    pub fn setup_fft<F: IsTwoAdicField>(
+        &self,
+        kernel: &str,
+        twiddles: &[FieldElement<F>],
+    ) -> Result<(&metal::CommandBufferRef, &metal::ComputeCommandEncoderRef), FFTError> {
+        let butterfly_kernel = self
+            .library
+            .get_function(kernel, None)
+            .map_err(FFTError::MetalFunctionError)?;
+
+        let pipeline = self
+            .device
+            .new_compute_pipeline_state_with_function(&butterfly_kernel)
+            .map_err(FFTError::MetalPipelineError)?;
+
+        let twiddles_buffer = {
+            let twiddles: Vec<_> = twiddles.iter().map(|elem| elem.value().clone()).collect();
+            let basetype_size = std::mem::size_of::<F::BaseType>();
+
+            self.device.new_buffer_with_data(
+                unsafe { mem::transmute(twiddles.as_ptr()) }, // reinterprets into a void pointer
+                (twiddles.len() * basetype_size) as u64,
+                MTLResourceOptions::StorageModeShared,
+            )
+        };
+
+        let command_buffer = self.command_queue.new_command_buffer();
+        let compute_encoder = command_buffer.new_compute_command_encoder();
+
+        compute_encoder.set_compute_pipeline_state(&pipeline);
+        compute_encoder.set_buffer(1, Some(&twiddles_buffer), 0);
+
+        Ok((command_buffer, compute_encoder))
+    }
+
+    pub fn execute_fft<F: IsTwoAdicField>(
+        &self,
+        input: &[FieldElement<F>],
+        (command_buffer, compute_encoder): (
+            &metal::CommandBufferRef,
+            &metal::ComputeCommandEncoderRef,
+        ),
+    ) -> Result<Vec<FieldElement<F>>, FFTError> {
+        let order = input.len().trailing_zeros() as u64;
```

Comment:
Casts may panic, we could file an issue for replacing these `as` conversions with a new `FFTError`.

Reply:
This way we ensure that if the code panics it'd be only because the `unsafe` block.

Reply:
This as won't panic, it's good, it's an u32 casted as an u64. Also, if it were the other way around, it wouldn't panic, it slices the value.

Reply:
TIL, I agree, the main drive of these casts is because the `metal` crate seems to only handle `u64` integers, I assumed that a `usize` couldn't be bigger than 64 bits in most hardware but in any way we must change them.